plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -132,6 +130,323 @@ nonZeroArg :: (Integer -> Integer -> Integer) -> Integer -> Integer -> Evaluatio
+
+5. A built-in function can fail. Whenever a builtin fails, evaluation of the whole program fails.
+There's a number of ways a builtin can fail:
+
+- as we've just seen a type conversion can fail due to an unsuccessful bounds check
+- if the builtin expects, say, a 'Text' argument, but gets fed an 'Integer' argument
+- if the builtin expects any constant, but gets fed a non-constant
+- if its denotation runs in the 'EvaluationResult' and an 'EvaluationFailure' gets returned
+
+Most of these are not a concern to the user defining a built-in function (conversions are handled
+within the builtin application machinery, type mismatches are on the type checker and the person
+writing the program etc), however explicitly returning 'EvaluationFailure' from a builtin is
+something that happens commonly.
+
+One simple example is a monomorphic function matching on a certain constructor and failing in all
+other cases:
+
+    toBuiltinMeaning UnIData =
+        makeBuiltinMeaning
+            (\case
+                I i -> EvaluationSuccess i
+                _   -> EvaluationFailure)
+            <costingFunction>
+
+The inferred type of the denotation is
+
+    Data -> EvaluationResult Integer
+
+and the Plutus type of the builtin is
+
+    data -> integer
+
+because the error effect is implicit in Plutus.
+
+Returning @EvaluationResult a@ for a type variable @a@ is also fine, i.e. it doesn't matter whether
+the denotation is monomorphic or polymorphic w.r.t. failing.
+
+But note that
+
+    'EvaluationResult' MUST BE EXPLICIYLY USED FOR ANY FAILING BUILTIN AND THROWING AN EXCEPTION
+    VIA 'error' OR 'throw' OR ELSE IS NOT ALLOWED AND CAN BE A HUGE VULNERABILITY. MAKE SURE THAT
+    NONE OF THE FUNCTIONS THAT YOU USE TO DEFINE A BUILTIN THROW EXCEPTIONS
+
+An argument of a builtin can't have 'EvaluationResult' in its type -- only the result.
+
+6. A builtin can emit log messages. For that it needs to run in the 'Emitter' monad. The ergonomics
+are the same as with 'EvaluationResult': 'Emitter' can't appear in the type of an argument and
+polymorphism is fine. For example:
+
+    toBuiltinMeaning Trace =
+        makeBuiltinMeaning
+            (\text a -> a <$ emitM text)
+            <costingFunction>
+
+The inferred type of the denotation is
+
+    forall a. Text -> a -> Emitter a
+
+and the Plutus type of the builtin is
+
+    all a. text -> a -> a
+
+because just like with the error effect, whether a function logs anything or not is not reflected
+in its type.
+
+'makeBuiltinMeaning' allows one to nest 'EvaluationResult' inside of 'Emitter' and vice versa,
+but as always nesting monads inside of each other without using monad transformers doesn't have good
+ergonomics, since computations of such a type can't be chained with a simple @(>>=)@.
+
+This concludes the list of simple cases. Before we jump to the hard ones, we need to talk about how
+polymorphism gets elaborated, so read Note [Elaboration of polymorphism] next.
```

Comment:
The reason why I created this barrier between the simple cases and complicated ones is because in the simple cases it's really trivial to add a builtin, you don't need any knowledge of the elaboration machinery, how things are represented etc -- you just write the most natural thing and it works. So pretty much anyone can do it and the simple cases cover the majority of usages. But for complicated cases you really have to take some deep dives into all this stuff and so only a trained individual can handle them. Hence I think it's worth distinguishing between "this is really simple, you don't need any extra knowledge and it'll be sufficient in the majority of cases" and "here's a shitload of material to learn before you attempt to do anything of this kind".